### ISSUE TITLE
Attribute Error Regarding MultiAgentObservationSpace

### ISSUE THREAD

**Issue by xijia-tao** [wwwCielwww/MADDPG#3]
When running `main.py`, I got

`C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\gym\logger.py:30: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize('%s: %s'%('WARN', msg % args), 'yellow'))
Traceback (most recent call last):
  File "main.py", line 9, in <module>
    model = maddpg(ma_policy, "PongDuel-v0")
  File "C:\Users\Ciel Tao\desktop\maddpg\maddpg\maddpg.py", line 98, in __init__
    n_episodes_rollout = td3_parameters.N_EPISODES_ROLLOUT)
  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_baselines3\td3\td3.py", line 111, in __init__
    supported_action_spaces=(gym.spaces.Box),
  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_baselines3\common\off_policy_algorithm.py", line 122, in __init__
    supported_action_spaces=supported_action_spaces,
  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_basel  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_baselines3\common\base_class.py", line 196, in _wrap_env
    env = DummyVecEnv([lambda: env])
  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_baselines3\common\vec_env\dummy_vec_env.py", line 29, in __init__
    self.keys, shapes, dtypes = obs_space_info(obs_space)
  File "C:\Users\Ciel Tao\.conda\envs\tensorflow_env\lib\site-packages\stable_baselines3-0.11.0a7-py3.6.egg\stable_baselines3\common\vec_env\util.py", line 72, in obs_space_info
    shapes[key] = box.shape
AttributeError: 'MultiAgentObservationSpace' object has no attribute 'shape'`

I suspect that the error is originated from the improper use of **stable baselines 3 algorithms** per se, which are intended for environments with a single agent. But the MultiAgentObservationSpace of `ma_gym` is in the form of a list, which contains single-agent observation space as elements.

**Developers**
@wwwCielwww @EE-LiuYunhao

**Comment by EE-LiuYunhao** [comment]
I saw the error was traced up towards the __init__ method. Therefore, I 
am thinking the error might can be resolved by changing some parameters
in the constructor method. Yet, I was not able to locate the particular
lines leading to the exception in my version of stable-baseline 3
library. I will double check the version to make sure that we are using
the same baseline codes.

**Comment by EE-LiuYunhao** [comment]
_I believe I have located the problem, desipte the differences between
the implementation of stable-baseline-3 lib across Win and Linux
platforms._

---

The problem is the stable-baseline-3 expects the input
environment presents the algorithm an ndarray or a pyTorch.Tensor
as the observation space. However, the ma-gym implementation takes
a simplified ﻿version, i.e. using the python List as the base for
the observation space. The very codes in ma-gym is pasted as
follows: 

```py
class MultiAgentObservationSpace(list):
    def __init__(self, agents_observation_space):
        for x in agents_observation_space:
            assert isinstance(x, gym.spaces.space.Space)

        super().__init__(agents_observation_space)
        self._agents_observation_space = agents_observation_space
```

while the following lines in the BasePolicy from stable-baseline-3
utilize the `shape` attribute. 

```py
def obs_space_info(obs_space):
    if isinstance(obs_space, gym.spaces.Dict):
        assert isinstance(obs_space.spaces, OrderedDict), "Dict space must have ordered subspaces"
        subspaces = obs_space.spaces
    elif isinstance(obs_space, gym.spaces.Tuple):
        subspaces = {i: space for i, space in enumerate(obs_space.spaces)}
    else:
        assert not hasattr(obs_space, "spaces"), f"Unsupported structured space '{type(obs_space)}'"
        subspaces = {None: obs_space}
    keys = []
    shapes = {}
    dtypes = {}
    for key, box in subspaces.items():
        keys.append(key)
        shapes[key] = box.shape 
        dtypes[key] = box.dtype
    return keys, shapes, dtypes
```

Note, since the obs_space from Ma-Gym is neither a `gym.space.spaces`
nor a `gym.space.tuple`, in the loop over `subspace.items()`, the
box will be just the `MultiAgentObservationSpace` itself
which is inherited from a List. Consequently, error will be
resulted in. 

As a conclusion, the ma-gym is not compatible with the 
stable-baseline-3, not because its multi-agent setting, but due
to the way they implement critical elements, one being based on 
native python list object while the other requiring numpy or
pytorch classes. **If we insite on the stable-baseline-3, we
have to RE-implement the entire environment.**

@Ericcsr Time for you to make a decision.

**Comment by xijia-tao** [comment]
> As a conclusion, the ma-gym is not compatible with the
> stable-baseline-3, not because its multi-agent setting, but due
> to the way they implement critical elements, one being based on
> native python list object while the other requiring numpy or
> pytorch classes. **If we insite on the stable-baseline-3, we have to RE-implement the entire environment.**

After reviewing the code you mentioned, I believe the problem is not about the implementation of `ma-gym` library. 

Frankly, if we are to create a multi-agent environment ourselves, the most straightforward way to do this is to pack `n` single-agent-like environments into a list (where `n` is the number of agents). *And those single-agent-like environments are exactly what the `obs_space_info` function you mentioned can process.* 

To illustrate my point, see the [notebook](https://github.com/wwwCielwww/MADDPG/blob/debug/demo/obs_space_info.ipynb).

In conclusion, I think it might be better to modify the stable-baselines3 code for our use, instead of directly calling them from a library.

**Comment by EE-LiuYunhao** [comment]
You are correct. However, to due to the inheritance relationship, modifying the `TD3` implementation might requires to
change the very bottom layer of the codes, i.e. the base policy. I am not sure whether the amount of work will be too
huge. 

@Ericcrs Eric, how about your opinion?

**Comment by Ericcsr** [comment]
Could we simple write a processor which take the list as input and convert it to be numpy or pytorch? I think numpy and pytorch support build data from list right?

**Comment by xijia-tao** [comment]
> Could we simple write a processor which take the list as input and convert it to be numpy or pytorch? I think numpy and pytorch support build data from list right?

Yeah, but that's not the point. 

For the 2-agent PongDuel environment:
```
>>> env = gym.make("PongDuel-v0")
>>> env.observation_space
[Box(0.0, 1.0, (10,), float32), Box(0.0, 1.0, (10,), float32)]
>>> env.action_space
[Discrete(3), Discrete(3)]
```
That's the list we're talking about. We obviously cannot change the list here into numpy.ndarray or tensor.

Having this noted, could you please offer us some suggestions for change? @Ericcsr

**Comment by EE-LiuYunhao** [comment]
- [ ] Review the original TD3 implementation to ensure the shapes required from the Env
       @EE-LiuYunhao 
- [ ] Create a wrapper to ensure the Env matches with the TD3 implementation
- [ ] Modify the actor-critic part based on the new Env, with TD3 origin actor or critic implementation
       @EE-LiuYunhao

**Comment by xijia-tao** [comment]
Fixed the problem by processing the `ma-gym` environment with the [`env_wrapper`](https://github.com/wwwCielwww/MADDPG/blob/master/hyper_parameters.py#L33) class in `hyper_parameters.py` as @EE-LiuYunhao and @Ericcsr suggested.

I'll close the issue now. We can reopen it in case of future issues :)
